### PR TITLE
Refactored build-metadata and intialization

### PR DESCRIPTION
**Pull request by heueristik**
The following scheme

For `prepareInstallation` and `prepareUninstallation`, we can
indicates that no function has to be called by writing "" for the `functionToCall` field.
```json
{
      "description": "",
      "inputs": []
}
```
If, for example, an initializer should be called, we write:
```json
{
      "description": "initialize build 1",
      "inputs": [{...}, {...}]
}
```

For `prepareUpdate`, every build that updates can transition from to the current version has a separate object, e.g.,
```json
"prepareUpdate": {
  "1": {
    "description": "initialize from build 1",
    "inputs": [{...}]
  }
  "2": {
    "description": "initialize from build 2",
    "inputs": []
  }
}
```
If there is no other build from which an update is supported (the initial build of an `PluginUUPSUpgradeable` or `Plugin` or `PluginCloneable`, we leave it empty
```json
"prepareUpdate": {}
```

Task: [OS-414](https://aragonassociation.atlassian.net/browse/OS-414)

[OS-414]: https://aragonassociation.atlassian.net/browse/OS-414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/release1/build1/build-metadata-R1B1.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "ui": "",
+  "change": "Initial build.",
+  "pluginSetupABI": {
+    "prepareInstallation": {
+      "functionToCall": "initialzeBuild1",
```
Comment (Michael-A-Heuer) [inline contracts/release1/build1/build-metadata-R1B1.json:6]: Would just "functionSelector" or "selector" be better?

Comment (brickpop) [inline contracts/release1/build1/build-metadata-R1B1.json]: Shouldn't this function be called by the PluginSetup? Why is it needed here?

File: contracts/release1/build1/build-metadata-R1B1.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      "arguments": ["uint256 number"]
+    },
+    "prepareUpdate": {},
+    "prepareUninstallation": {
+      "functionToCall": "",
```
Comment (brickpop) [inline contracts/release1/build1/build-metadata-R1B1.json]: Why do we need this? It is not being used.

If we are to use this, I would suggest `functionName` instead. 
Should we pass the full function definition instead of just the name?

File: test/integration-testing/simple-storage-setup-processing.ts [hunk 3]
```diff
@@ -348,7 +348,7 @@ describe('SimpleStorage Integration', function () {
           pluginSetupRefR1B1,
           pluginSetupRefR1B3,
           ethers.utils.defaultAbiCoder.encode(
-            buildMetadata2.pluginSetupABI.prepareUpdate.fromBuild1,
+            buildMetadata2.pluginSetupABI.prepareUpdate[1].arguments,
```
Comment (brickpop) [inline test/integration-testing/simple-storage-setup-processing.ts]: Technically, the current JSON should be accessed like `["1"]` instead of `[1]`, despite it works

File: contracts/release1/build1/build-metadata-R1B1.json (hunk 1 above)
Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json]: Right, I thought that we can get more information from the NatSpec of the function that is being called, but you are right - the call happens in the `PluginSetup` and more information can be given as input that might be needed in the setup
Instead, we might want to have an array of descriptions for each input argument.

Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json]: I chose the same JSON structure now that is also used in the contract ABI

For comparison, here the contract ABI of `createProposal` from `TokenVoting`:
```json
    {
      "inputs": [
        {
          "internalType": "bytes",
          "name": "_metadata",
          "type": "bytes"
        },
        {
          "components": [
            {
              "internalType": "address",
              "name": "to",
              "type": "address"
            },
            {
              "internalType": "uint256",
              "name": "value",
              "type": "uint256"
            },
            {
              "internalType": "bytes",
              "name": "data",
              "type": "bytes"
            }
          ],
          "internalType": "struct IDAO.Action[]",
          "name": "_actions",
          "type": "tuple[]"
        },
        {
          "internalType": "uint256",
          "name": "_allowFailureMap",
          "type": "uint256"
        },
        {
          "internalType": "bool",
          "name": "_approveProposal",
          "type": "bool"
        },
        {
          "internalType": "bool",
          "name": "_tryExecution",
          "type": "bool"
        },
        {
          "internalType": "uint64",
          "name": "_startDate",
          "type": "uint64"
        },
        {
          "internalType": "uint64",
          "name": "_endDate",
          "type": "uint64"
        }
      ],
      "name": "createProposal",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "proposalId",
          "type": "uint256"
        }
      ],
      "stateMutability": "nonpayable",
      "type": "function"
    },
```
```

File: contracts/release1/build1/build-metadata-R1B1.json (hunk 2 above)
Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json]: See above

File: test/integration-testing/simple-storage-setup-processing.ts (hunk 3 above)
Comment (heueristik) [reply test/integration-testing/simple-storage-setup-processing.ts]: Fixed in bdbe2936c70eeb609617939b39959806a64bd7e3

File: contracts/release1/build1/build-metadata-R1B1.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      "description": "The information required for the installation of build 1.",
+      "inputs": [
+        {
+          "name": "number",
+          "type": "uint256",
+          "description": "The inital number to be stored."
+        }
+      ]
+    },
+    "prepareUpdate": {},
```
Comment (Rekard0) [inline contracts/release1/build1/build-metadata-R1B1.json:16]: why not also add description & inputs here? specially description, expaining why there is no inputs.

Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json:16]: My thinking was that the absence of objects inside `"prepareUpdate": {}` indicates that there are no possible updates. 
If there are builds from which an update is possible, we add them as an object.

Let's imagine we write the metadata for build 5
```json
    "prepareUpdate": {
      "1": {
        "description": "The information required for the update from build 1 to build 5.",
        "inputs": [
          {
            "name": "account",
            "type": "address",
            "description": "The inital account to be stored."
          }
        ]
      },
      "2": {
        //...
      },
      "4": {
        //...
      }
    },
```
In this example, you, for some reason, cannot upgrade from build 3 to build 5.

Do you think we are missing a top-level description for all the updates?

File: contracts/release1/build1/build-metadata-R1B1.json (hunk 2 above)
Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json]: We moved to a new format so this can be resolved

File: contracts/release1/build1/build-metadata-R1B1.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
+{
+  "ui": "",
```
Comment (brickpop) [inline contracts/release1/build1/build-metadata-R1B1.json]: Same as on the other PR, I recommend leaving this field as an object for future proofness. I can see many keys being part of the UI section

File: contracts/release1/build2/build-metadata-R1B2.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
+{
+  "ui": "",
+  "change": "Added `address account` as a second state variable.",
+  "pluginSetupABI": {
```
Comment (brickpop) [inline contracts/release1/build2/build-metadata-R1B2.json]: Wondering if there's a better name for this
- `pluginSetup`
- `lifecycleABI`
- `lifecycleParams`
- ...

File: contracts/release1/build3/build-metadata-R1B3.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
+{
+  "ui": "",
+  "change": "Replaced the `STORE_PERMISSION` permission by the more specific `STORE_NUMBER_PERMISSION` and `STORE_ACCOUNT_PERMISSION` and added the `NumberStored` and `AccountStored` events.",
```
Comment (brickpop) [inline contracts/release1/build3/build-metadata-R1B3.json]: I know this is just an example, but I would encourage developers to add human readable descriptions by giving an example ourselves

File: test/unit-testing/simple-stroage-r1b3-setup.ts [hunk 8]
```diff
@@ -159,7 +161,9 @@ describe('SimpleStorageR1B3Setup', function () {
               plugin: pluginBuild1.address,
               currentHelpers: [],
               data: ethers.utils.defaultAbiCoder.encode(
-                buildMetadata3.pluginSetupABI.prepareUpdate.fromBuild1,
+                buildMetadata3.pluginSetupABI.prepareUpdate['1'].inputs.map(
```
Comment (brickpop) [inline test/unit-testing/simple-stroage-r1b3-setup.ts]: Either everything should be a string or a number. See the diff block below.

Speaking of JS/TS, it's probably better to rely on strings, since objects indexed by numbers may resemble an array and open the door to unexpected edge cases

File: test/unit-testing/simple-stroage-r1b3-setup.ts [hunk 9]
```diff
@@ -218,7 +222,9 @@ describe('SimpleStorageR1B3Setup', function () {
               plugin: pluginBuild2.address,
               currentHelpers: [],
               data: ethers.utils.defaultAbiCoder.encode(
-                buildMetadata3.pluginSetupABI.prepareUpdate.fromBuild2,
+                buildMetadata3.pluginSetupABI.prepareUpdate[2].inputs.map(
```
Comment (brickpop) [inline test/unit-testing/simple-stroage-r1b3-setup.ts]: See the comment above

File: contracts/release1/build1/build-metadata-R1B1.json (hunk 5 above)
Comment (heueristik) [reply contracts/release1/build1/build-metadata-R1B1.json]: Changed in 53317c7760d8569368a1f30470367eaaba7b803d

File: contracts/release1/build2/build-metadata-R1B2.json (hunk 6 above)
Comment (heueristik) [reply contracts/release1/build2/build-metadata-R1B2.json]: The functions 
`prepareInstallation`, `prepareUpdate`, `prepareUninstallation` that require the arguments that we describe with types and names are located in the `PluginSetup` contract.
Thus, naming this the `pluginSetup` is good, and I would not refrain from that. 

Since what we write there is not exactly the ABI (the contract ABI is a standardized way of encoding its functions and parameters into a format that can be understood by other programs or contracts) I have nothing against dropping the `ABI` part from the name.

In this context, talking about lifecyles makes things more vague IMO.

WDYT @Rekard0 ?

File: test/unit-testing/simple-stroage-r1b3-setup.ts (hunk 8 above)
Comment (heueristik) [reply test/unit-testing/simple-stroage-r1b3-setup.ts]: fae5fab68683ad7b63caa38dc164d959b3d92577

File: test/unit-testing/simple-stroage-r1b3-setup.ts (hunk 9 above)
Comment (heueristik) [reply test/unit-testing/simple-stroage-r1b3-setup.ts]: fae5fab68683ad7b63caa38dc164d959b3d92577

File: contracts/release1/build2/build-metadata-R1B2.json (hunk 6 above)
Comment (heueristik) [reply contracts/release1/build2/build-metadata-R1B2.json]: Changed in aff41b9b35e41c3b5745e9e8d3f820637112a858

File: contracts/release1/build3/build-metadata-R1B3.json (hunk 7 above)
Comment (heueristik) [reply contracts/release1/build3/build-metadata-R1B3.json]: Changed in 8d1269e32f811223b47a34c552eeddddeaa2a922

**Review by brickpop — APPROVED** [review]
